packages/http/src/mocker/negotiator/NegotiatorHelpers.ts

```diff
@@ -53,236 +60,283 @@ function findResponseByStatusCode(
-        };
-      } else {
-        throw ProblemJsonError.fromTemplate(
-          NOT_ACCEPTABLE,
-          `Could not find any content that satisfies ${mediaTypes.join(',')}`,
-        );
+            mediaType: 'text/plain',
+            headers: headers || [],
+          });
+        }
       }
-    }
-    // user did not provide mediaType
-    // OR
-    // a httpContent for a provided mediaType does not exist
-    return helpers.negotiateDefaultMediaType(
-      {
-        code,
-        dynamic,
-        exampleKey,
-      },
-      response,
-    );
+      // user did not provide mediaType
+      // OR
+      // a httpContent for a provided mediaType does not exist
+      logger.trace('No mediaType provided. Fallbacking to the default media type (application/json)');
+      return helpers.negotiateDefaultMediaType(
+        {
+          code,
+          dynamic,
+          exampleKey,
+        },
+        response,
+      );
+    });
   },
 
   negotiateOptionsForDefaultCode(
     httpOperation: IHttpOperation,
     desiredOptions: NegotiationOptions,
-  ): IHttpNegotiationResult {
+  ): Reader<Logger, Either<Error, IHttpNegotiationResult>> {
     const lowest2xxResponse = findLowest2xx(httpOperation.responses);
     if (lowest2xxResponse) {
       return helpers.negotiateOptionsBySpecificResponse(httpOperation, desiredOptions, lowest2xxResponse);
     }
-    throw new Error('No 2** response defined, cannot mock');
+
+    return reader.of(left(new Error('No 2** response defined, cannot mock')));
   },
 
   negotiateOptionsBySpecificCode(
     httpOperation: IHttpOperation,
     desiredOptions: NegotiationOptions,
     code: string,
-  ): IHttpNegotiationResult {
+  ): Reader<Logger, Either<Error, IHttpNegotiationResult>> {
     // find response by provided status code
-    const responseByForcedStatusCode =
-      findResponseByStatusCode(httpOperation.responses, code) ||
-      createResponseFromDefault(httpOperation.responses, code);
-    if (responseByForcedStatusCode) {
-      try {
-        // try to negotiate
-        return helpers.negotiateOptionsBySpecificResponse(httpOperation, desiredOptions, responseByForcedStatusCode);
-      } catch (error) {
-        // if negotiations fail try a default code
-        try {
-          return helpers.negotiateOptionsForDefaultCode(httpOperation, desiredOptions);
-        } catch (error2) {
-          throw new Error(`${error}. We tried default response, but we got ${error2}`);
-        }
+    return withLogger(logger => {
+      const result = findResponseByStatusCode(httpOperation.responses, code);
+      if (!result) {
+        logger.info(`Unable to find a ${code} response definition`);
+        return createResponseFromDefault(httpOperation.responses, code);
       }
-    }
-    // if no response found under a status code throw an error
-    throw new Error('Requested status code is not defined in the schema.');
+
+      return result;
+    }).chain(responseByForcedStatusCode => {
```

Comment:
What's the benefit of using chain vs if?

Reply:
`if` and `chain` do two different things, they cannot be interchanged.

To keep it short, `chain` is just like a `flatMap`

`[1,2,3,4].map(t=>t*2) = [2,4,6,8]`
`[1,2,3].flatMap(t=>[t, t * 2, t * 3]) = [1, 2, 3, 2, 4, 6, 3, 6, 9]`

---

In this context in particular, `chain` unwraps the effect for you so if you have an Either — you can concatenate it with another Either

```ts

const operationThatMightFail = getDataFromInternet();

operationThatMightFail
  .map(operation => operation.result)
  .chain(result => postInternetDataSomewhere(result))
```

In this case you can see:

1. You receive an `Either` from a function that gets data from the internet or fails.
2. In case it passes, it will call the `map` function that will effectively transform the result
3. Pass the data to chain; a chain function can return another `Either` that will be "executed"/"unwrapped" for you automatically, flattening the final Either you have to deal with.

Does that help?

Reply:
I never liked `chain` / `bind` names for it. `flatMap` is so obvious a name! :)